server/src/main/java/org/elasticsearch/env/NodeEnvironment.java

```diff
@@ -295,13 +308,111 @@ public NodeEnvironment(Settings settings, Environment environment) throws IOExce
     }
 
     /**
-     * Resolve a specific nodes/{node.id} path for the specified path and node lock id.
-     *
-     * @param path       the path
-     * @return the resolved path
+     * Upgrades all data paths that have been written to by an older ES version to the 8.0+ compatible folder layout,
+     * removing the "nodes/${lockId}" folder prefix
      */
-    public static Path resolveNodePath(final Path path) {
-        return path.resolve(NODES_FOLDER).resolve("0");
+    private static boolean upgradeLegacyNodeFolders(Logger logger, Environment environment, NodeLock nodeLock) throws IOException {
+        boolean upgradeNeeded = false;
+
+        // check if we can do an auto-upgrade
+        for (Path path : environment.dataFiles()) {
+            final Path nodesFolderPath = path.resolve("nodes");
+            if (Files.isDirectory(nodesFolderPath)) {
+                final List<Integer> nodeLockIds = new ArrayList<>();
+
+                try (DirectoryStream<Path> stream = Files.newDirectoryStream(nodesFolderPath)) {
+                    for (Path nodeLockIdPath : stream) {
+                        String fileName = nodeLockIdPath.getFileName().toString();
+                        if (Files.isDirectory(nodeLockIdPath) && fileName.chars().allMatch(Character::isDigit)) {
+                            int nodeLockId = Integer.parseInt(fileName);
+                            nodeLockIds.add(nodeLockId);
+                        }
+                    }
+                }
+
+                if (nodeLockIds.isEmpty() == false) {
+                    upgradeNeeded = true;
+
+                    if (nodeLockIds.equals(Arrays.asList(0)) == false) {
+                        throw new IllegalStateException("Cannot upgrade multiple path with multiple lock ids or lock id different to 0 " +
```

Comment:
I think we'll have users hitting this message and it's a bit cryptic so will raise questions. How about

```
data path [...] cannot be upgraded automatically because it contains data from nodes with ordinals [...], due to use of obsolete [node.max_local_storage_nodes] setting
```

Maybe we can also guide them to some docs?

Reply:
adapted message in e22c2ee